671 added search features

kronken opened the pull request:
Closes #671 

- Default search is like before
- Wildcard search will be enabled if search input contains '*', wildcard search makes the searching a lot slower.
- Showing short id of nodes in search list
- Order routes in list based on modified on

Divinare commented:
Testing:
- fix console warning
- after write, start search
- loading icon
- writing nonstop?
- when searching for 4562, sub group should be shorted by name

kronken commented:
Refactored to use delayed search. 
Search will be triggered when user hasn't written anything for 500 ms.
Also renamed lineSeach to searchInput, since we use it both for lines and nodes.
Also fixed console warnings, loading icon, nonstop writing, ordering by id

Divinare commented on src/stores/searchResultStore.ts:
```diff
@@ -33,34 +57,42 @@ export class SearchResultStore {
 
-    public getFilteredItems = (): (INodeBase | ILine)[] => {
+    @action
+    private searchAsync = async () => {
```
Remove current method search, rename this as search and put:
this.setIsSearching(true);
{searchAsync code}
this.setIsSearching(false);

kronken replied:
Hmm, yeah. The reason why I split it was since I thought the UI won't re-render if we don't have the search logic in a async function. But I tried this, and it seems to work equally well to just have everything in the same method. I can change this

kronken commented on src/components/shared/searchView/SearchResults.tsx:
```diff
@@ -92,15 +97,16 @@ class SearchResults extends React.Component<ISearchResultsProps, ISearchResultsS
         return this.props.searchResultStore!
-            .getFilteredItems()
+            .filteredItems
+            .slice()
```
Old code was throwing warnings in the console. Splice will alter the array in place, which is not allowed outside of an mobx action.

kronken commented on src/stores/searchResultStore.ts:
```diff
@@ -33,34 +57,42 @@ export class SearchResultStore {
 
-            // Filter by line.id
-            if (line.id.indexOf(searchInput) > -1) return true;
+    private matchText(text: string, rule: string) {
```
Could change it to searchInput. Regex usually use the word rule, but since this function doesn't include regex, searchInput would indeed be better

kronken commented:
Added trimming and fixed issue with no reloading after toggling line/node

Divinare commented:
Review & testing ok. Works nicely now!
